Adding a version serialization type to ConvPackedParam

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #43651 ConvPackedParams: remove legacy format
* **#43086 Adding a version serialization type to ConvPackedParam**
* #43524 quantized conv: add support for graph mode BC testing, and increase coverage

This PR changes the format of `ConvPackedParam` in a nearly backwards-compatible way:
* a new format is introduced which has more flexibility and a lower on-disk size
* custom pickle functions are added to `ConvPackedParams` which know how to load the old format
* the custom pickle functions are **not** BC because the output type of `__getstate__` has changed.  We expect this to be acceptable as no user flows are actually broken (loading a v1 model with v2 code works), which is why we whitelist the failure.

For a short forward compatibility period, we keep the `__getstate__` function with the old format,
while `__getstate__` knows about both old and new format.  The next PR in the stack will be landed
some period after this PR, and will switch `__getstate__` to the new format.

Note: the bc test cases were generated with the `v2` format, to allow manual testing of the `v2`
code in this PR.

Test plan:

```
// adhoc testing of saving v1 and loading in v2: https://gist.github.com/vkuzo/f3616c5de1b3109cb2a1f504feed69be 

// test that loading models with v1 conv params format works and leads to the same numerics
python test/test_quantization.py TestSerialization.test_conv2d_graph
python test/test_quantization.py TestSerialization.test_conv2d_nobias_graph

// test that saving and loading models with v2 conv params format works and leads to same numerics
python test/test_quantization.py TestSerialization.test_conv2d_graph_v2
python test/test_quantization.py TestSerialization.test_conv2d_nobias_graph_v2

// test that numerics on quantized MobileNetV2 do not change:
// https://gist.github.com/vkuzo/d593385a88c30e22f46eba4248070c79
```

Note: this is a newer copy of https://github.com/pytorch/pytorch/pull/40003

Differential Revision: [D23347832](https://our.internmc.facebook.com/intern/diff/D23347832)

vkuzo committed:
Adding a version serialization type to ConvPackedParam

[ghstack-poisoned]

vkuzo committed:
Update on "not for land: Adding a version serialization type to ConvPackedParam"


Note: this is a debug copy of https://github.com/pytorch/pytorch/pull/40003

[ghstack-poisoned]

vkuzo committed:
Update on "not for land: Adding a version serialization type to ConvPackedParam"


Note: this is a debug copy of https://github.com/pytorch/pytorch/pull/40003

[ghstack-poisoned]

vkuzo committed:
Update on "not for land: Adding a version serialization type to ConvPackedParam"


Note: this is a debug copy of https://github.com/pytorch/pytorch/pull/40003

[ghstack-poisoned]

vkuzo committed:
Update on "not for land: Adding a version serialization type to ConvPackedParam"


Note: this is a debug copy of https://github.com/pytorch/pytorch/pull/40003

[ghstack-poisoned]

vkuzo committed:
Update on "not for land: Adding a version serialization type to ConvPackedParam"


Note: this is a debug copy of https://github.com/pytorch/pytorch/pull/40003

[ghstack-poisoned]

vkuzo committed:
Update on "not for land: Adding a version serialization type to ConvPackedParam"


Note: this is a debug copy of https://github.com/pytorch/pytorch/pull/40003

[ghstack-poisoned]

vkuzo committed:
Update on "not for land: Adding a version serialization type to ConvPackedParam"


Note: this is a debug copy of https://github.com/pytorch/pytorch/pull/40003

[ghstack-poisoned]

vkuzo committed:
Update on "not for land: Adding a version serialization type to ConvPackedParam"


Note: this is a debug copy of https://github.com/pytorch/pytorch/pull/40003

[ghstack-poisoned]

vkuzo committed:
Update on "not for land: Adding a version serialization type to ConvPackedParam"


Note: this is a debug copy of https://github.com/pytorch/pytorch/pull/40003

[ghstack-poisoned]

vkuzo committed:
Update on "not for land: Adding a version serialization type to ConvPackedParam"


Note: this is a debug copy of https://github.com/pytorch/pytorch/pull/40003

[ghstack-poisoned]

vkuzo committed:
Update on "not for land: Adding a version serialization type to ConvPackedParam"


Note: this is a debug copy of https://github.com/pytorch/pytorch/pull/40003

[ghstack-poisoned]

vkuzo committed:
Update on "not for land: Adding a version serialization type to ConvPackedParam"


Note: this is a debug copy of https://github.com/pytorch/pytorch/pull/40003

[ghstack-poisoned]

vkuzo commented on torch/custom_class.h:
```diff
@@ -207,16 +224,19 @@ class class_ {
-        " but expected ",
-        ser_type->repr_str());
+
+    if (validate_setstate_input_type) {
```
(moving over @dzhulgakov comment from https://github.com/pytorch/pytorch/pull/43230)

if we want to allow IValue setter officially, we could also not have a separate split between def_pickle and def_pickle_do_not_use, just have a single one and allow the setstate argument to be Any. You can be even fancier and check isSubType on the type to just relax condition (i.e. that the setstate's argument type should be same or broader than the getstate's). I think it's pretty clean

I actually don't see a big harm in allowing the set

vkuzo replied:
Sure.  One other thing to think about is whether we are (1) checking this in for the long term or (2) as a hack to be deleted ASAP after VIP customers migrate.  Currently it's following the second, but we can switch to the first if that's the consensus.

dzhulgakov replied:
Let's go for long term (in which case we should modify the docs too). The right thing then is to change `==` into `isSubTypeOf` below. It'd work for IValue (which has type `Any`) but would also work for some more narrow relaxations. Seems like our offline discussion converged to that too, let's do it

dzhulgakov commented on schemas.txt:
```diff
@@ -0,0 +1,2360 @@
+aten::as_tensor.float(float t, *, int? dtype=None, Device? device=None) -> (Tensor)
```
looks like you included a random file

vkuzo replied:
sorry about that, will remove

dzhulgakov commented on test/backward_compatibility/check_backward_compatibility.py:
```diff
@@ -60,6 +60,9 @@
     ("aten::_compute_linear_combination", datetime.date(2020, 9, 1)),
+    # TODO before land: make the matching logic in this test more generic
+    #   to allow matching by function arguments
+    ("getstate", datetime.date(2020, 9, 1)),
```
why do you need it? and why is it named without the prefix?

vkuzo replied:
> why do you need it?

Afaik changing the return type of `__getstate__` will break this test as written, because it checks for function signatures.  We are saying (and I'd love feedback if this is incorrect) that this breakage is OK because all the user flows in Quantization callsits will still work via our patch, and we are OK with breaking non-Quantization code which depended on the old return type (we don't know of any real instances of this).

> why is it named without the prefix?

should be fixed after the TODO is implemented, we need to update this test a bit to also allow matching on arguments (currently it just matches on `__getstate__` which would match too many things).

dzhulgakov replied:
makes sense, but I don't think you need matching on arguments. We return the name of the same as a non-qualified name right now and it doesn't make much sense. We could just change code here (https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/core/custom_class.cpp#L49) to return `fn->qualname` as the name for the schema

or as you pointed out, we could post process it in python and just prepend `.arguments[0].type.name()` to the name (assuming that `self` is always the first argument)

vkuzo committed:
Update on "Adding a version serialization type to ConvPackedParam"


This PR changes the format of `ConvPackedParam` in a nearly backwards-compatible way:
* a new format is introduced which has more flexibility and a lower on-disk size
* custom pickle functions are added to `ConvPackedParams` which know how to load the old format
* the custom pickle functions are **not** BC because the output type of `__getstate__` has changed.  We expect this to be acceptable as no user flows are actually broken (loading a v1 model with v2 code works), which is why we whitelist the failure.

Test plan (TODO finalize):
* verify that saving and loading models with https://gist.github.com/vkuzo/f3616c5de1b3109cb2a1f504feed69be works
* TODO before land: add test coverage for loading v1 models
* TODO before land: verify numerics do not change
* TODO before land: verify all combinations of (trace, script) x (bias, no_bias) work


Note: this is a newer copy of https://github.com/pytorch/pytorch/pull/40003

[ghstack-poisoned]

vkuzo committed:
improve BC tests on "Adding a version serialization type to ConvPackedParam"


This PR changes the format of `ConvPackedParam` in a nearly backwards-compatible way:
* a new format is introduced which has more flexibility and a lower on-disk size
* custom pickle functions are added to `ConvPackedParams` which know how to load the old format
* the custom pickle functions are **not** BC because the output type of `__getstate__` has changed.  We expect this to be acceptable as no user flows are actually broken (loading a v1 model with v2 code works), which is why we whitelist the failure.

Test plan (TODO finalize):
* verify that saving and loading models with https://gist.github.com/vkuzo/f3616c5de1b3109cb2a1f504feed69be works
* TODO before land: add test coverage for loading v1 models
* TODO before land: verify numerics do not change
* TODO before land: verify all combinations of (trace, script) x (bias, no_bias) work


Note: this is a newer copy of https://github.com/pytorch/pytorch/pull/40003

[ghstack-poisoned]

vkuzo committed:
Update on "Adding a version serialization type to ConvPackedParam"


This PR changes the format of `ConvPackedParam` in a nearly backwards-compatible way:
* a new format is introduced which has more flexibility and a lower on-disk size
* custom pickle functions are added to `ConvPackedParams` which know how to load the old format
* the custom pickle functions are **not** BC because the output type of `__getstate__` has changed.  We expect this to be acceptable as no user flows are actually broken (loading a v1 model with v2 code works), which is why we whitelist the failure.

Test plan (TODO finalize):
* verify that saving and loading models with https://gist.github.com/vkuzo/f3616c5de1b3109cb2a1f504feed69be works
* TODO before land: add test coverage for loading v1 models
* TODO before land: verify numerics do not change
* TODO before land: verify all combinations of (trace, script) x (bias, no_bias) work


Note: this is a newer copy of https://github.com/pytorch/pytorch/pull/40003

[ghstack-poisoned]

vkuzo committed:
Update on "Adding a version serialization type to ConvPackedParam"


This PR changes the format of `ConvPackedParam` in a nearly backwards-compatible way:
* a new format is introduced which has more flexibility and a lower on-disk size
* custom pickle functions are added to `ConvPackedParams` which know how to load the old format
* the custom pickle functions are **not** BC because the output type of `__getstate__` has changed.  We expect this to be acceptable as no user flows are actually broken (loading a v1 model with v2 code works), which is why we whitelist the failure.

Test plan (TODO finalize):

```
// adhoc testing of saving v1 and loading in v2: https://gist.github.com/vkuzo/f3616c5de1b3109cb2a1f504feed69be 

// test that loading models with v1 conv params format works and leads to the same numerics
python test/test_quantization.py TestSerialization.test_conv2d_graph
python test/test_quantization.py TestSerialization.test_conv2d_nobias_graph

// test that saving and loading models with v2 conv params format works and leads to same numerics
python test/test_quantization.py TestSerialization.test_conv2d_graph_v2
python test/test_quantization.py TestSerialization.test_conv2d_nobias_graph_v2

// TODO before land: 
// test numerics for a real model
// test legacy ONNX path
```

Note: this is a newer copy of https://github.com/pytorch/pytorch/pull/40003

[ghstack-poisoned]

vkuzo committed:
Update on "Adding a version serialization type to ConvPackedParam"


This PR changes the format of `ConvPackedParam` in a nearly backwards-compatible way:
* a new format is introduced which has more flexibility and a lower on-disk size
* custom pickle functions are added to `ConvPackedParams` which know how to load the old format
* the custom pickle functions are **not** BC because the output type of `__getstate__` has changed.  We expect this to be acceptable as no user flows are actually broken (loading a v1 model with v2 code works), which is why we whitelist the failure.

Test plan (TODO finalize):

```
// adhoc testing of saving v1 and loading in v2: https://gist.github.com/vkuzo/f3616c5de1b3109cb2a1f504feed69be 

// test that loading models with v1 conv params format works and leads to the same numerics
python test/test_quantization.py TestSerialization.test_conv2d_graph
python test/test_quantization.py TestSerialization.test_conv2d_nobias_graph

// test that saving and loading models with v2 conv params format works and leads to same numerics
python test/test_quantization.py TestSerialization.test_conv2d_graph_v2
python test/test_quantization.py TestSerialization.test_conv2d_nobias_graph_v2

// TODO before land: 
// test numerics for a real model
// test legacy ONNX path
```

Note: this is a newer copy of https://github.com/pytorch/pytorch/pull/40003

[ghstack-poisoned]

dreiss commented on aten/src/ATen/native/quantized/cpu/conv_serialization.h:
```diff
@@ -0,0 +1,248 @@
+ * - Fields:
+ *  0. version (string)
+ *  1. list of non-optional tensors
+ *    0: packed parameters (int64_t)
```
Can we make this an unspecified integer type, and have the deserializer handle any integer type, then serialize with int8?  That saves a few percent of model size on the model I tested.

dreiss commented on aten/src/ATen/native/quantized/cpu/conv_serialization.h:
```diff
@@ -0,0 +1,248 @@
+    }
+  }
+  TORCH_INTERNAL_ASSERT(version != -1, "Unable to parse serialization version");
+  TORCH_INTERNAL_ASSERT(version >= 1 && version <= 2, "Unknown serialization version");
```
I would skip this since the check above will already fire.  Just one more thing to update if we add another version.

dreiss commented on aten/src/ATen/native/quantized/cpu/conv_serialization.h:
```diff
@@ -0,0 +1,248 @@
+    state = std::tie(version, non_optional, optional);
+
+  } else {
+    // version 2
+
+    auto elements = v.toTuple()->elements();
+    std::string version = elements[0].toStringRef();
+    torch::List<at::Tensor> non_optional = elements[1].toTensorList();
+    // TODO before land: figure out if there is a better way to
+    //   cast to OptionalTensorList
+    torch::List<c10::IValue> optional = elements[2].toList();
+
+    std::vector<at::Tensor> non_optional_vec;
+    non_optional_vec.insert(non_optional_vec.end(), non_optional.begin(), non_optional.end());
+
+    std::vector<c10::optional<at::Tensor>> optional_vec;
+    c10::optional<at::Tensor> bias = optional.get(0).toOptional<at::Tensor>();
+    optional_vec.push_back(bias);
+
+    state = std::tie(version, non_optional_vec, optional_vec);
```
Does `return v.to<ConvParamsSerializationType>();` not work here?

dreiss commented on aten/src/ATen/native/quantized/cpu/conv_serialization.h:
```diff
@@ -0,0 +1,248 @@
+  c10::optional<at::Tensor> bias;
+  std::tie(weight, bias) = params->unpack();
+
+  non_optional.emplace_back(std::move(params_tensor));
+  non_optional.emplace_back(std::move(weight));
```
Why bother moving here?  Tensor is effectively a pointer.

dreiss reviewed (commented):
Looks pretty good to me.  This reduces the PT model size overhead on the model I tested from 54% to 24% (22 with int8 or int16 config params).

vkuzo committed:
comments on "Adding a version serialization type to ConvPackedParam"


This PR changes the format of `ConvPackedParam` in a nearly backwards-compatible way:
* a new format is introduced which has more flexibility and a lower on-disk size
* custom pickle functions are added to `ConvPackedParams` which know how to load the old format
* the custom pickle functions are **not** BC because the output type of `__getstate__` has changed.  We expect this to be acceptable as no user flows are actually broken (loading a v1 model with v2 code works), which is why we whitelist the failure.

Test plan (TODO finalize):

```
// adhoc testing of saving v1 and loading in v2: https://gist.github.com/vkuzo/f3616c5de1b3109cb2a1f504feed69be 

// test that loading models with v1 conv params format works and leads to the same numerics
python test/test_quantization.py TestSerialization.test_conv2d_graph
python test/test_quantization.py TestSerialization.test_conv2d_nobias_graph

// test that saving and loading models with v2 conv params format works and leads to same numerics
python test/test_quantization.py TestSerialization.test_conv2d_graph_v2
python test/test_quantization.py TestSerialization.test_conv2d_nobias_graph_v2

// TODO before land: 
// test numerics for a real model
// test legacy ONNX path
```

Note: this is a newer copy of https://github.com/pytorch/pytorch/pull/40003

Differential Revision: [D23347832](https://our.internmc.facebook.com/intern/diff/D23347832)

[ghstack-poisoned]

vkuzo committed:
rebase on "Adding a version serialization type to ConvPackedParam"


This PR changes the format of `ConvPackedParam` in a nearly backwards-compatible way:
* a new format is introduced which has more flexibility and a lower on-disk size
* custom pickle functions are added to `ConvPackedParams` which know how to load the old format
* the custom pickle functions are **not** BC because the output type of `__getstate__` has changed.  We expect this to be acceptable as no user flows are actually broken (loading a v1 model with v2 code works), which is why we whitelist the failure.

Test plan (TODO finalize):

```
// adhoc testing of saving v1 and loading in v2: https://gist.github.com/vkuzo/f3616c5de1b3109cb2a1f504feed69be 

// test that loading models with v1 conv params format works and leads to the same numerics
python test/test_quantization.py TestSerialization.test_conv2d_graph
python test/test_quantization.py TestSerialization.test_conv2d_nobias_graph

// test that saving and loading models with v2 conv params format works and leads to same numerics
python test/test_quantization.py TestSerialization.test_conv2d_graph_v2
python test/test_quantization.py TestSerialization.test_conv2d_nobias_graph_v2

// TODO before land: 
// test numerics for a real model
// test legacy ONNX path
```

Note: this is a newer copy of https://github.com/pytorch/pytorch/pull/40003

Differential Revision: [D23347832](https://our.internmc.facebook.com/intern/diff/D23347832)

[ghstack-poisoned]

vkuzo committed:
Update on "Adding a version serialization type to ConvPackedParam"


This PR changes the format of `ConvPackedParam` in a nearly backwards-compatible way:
* a new format is introduced which has more flexibility and a lower on-disk size
* custom pickle functions are added to `ConvPackedParams` which know how to load the old format
* the custom pickle functions are **not** BC because the output type of `__getstate__` has changed.  We expect this to be acceptable as no user flows are actually broken (loading a v1 model with v2 code works), which is why we whitelist the failure.

Test plan (TODO finalize):

```
// adhoc testing of saving v1 and loading in v2: https://gist.github.com/vkuzo/f3616c5de1b3109cb2a1f504feed69be 

// test that loading models with v1 conv params format works and leads to the same numerics
python test/test_quantization.py TestSerialization.test_conv2d_graph
python test/test_quantization.py TestSerialization.test_conv2d_nobias_graph

// test that saving and loading models with v2 conv params format works and leads to same numerics
python test/test_quantization.py TestSerialization.test_conv2d_graph_v2
python test/test_quantization.py TestSerialization.test_conv2d_nobias_graph_v2

// TODO before land: 
// test numerics for a real model
// test legacy ONNX path
```

Note: this is a newer copy of https://github.com/pytorch/pytorch/pull/40003

Differential Revision: [D23347832](https://our.internmc.facebook.com/intern/diff/D23347832)

[ghstack-poisoned]

vkuzo committed:
Update on "Adding a version serialization type to ConvPackedParam"


This PR changes the format of `ConvPackedParam` in a nearly backwards-compatible way:
* a new format is introduced which has more flexibility and a lower on-disk size
* custom pickle functions are added to `ConvPackedParams` which know how to load the old format
* the custom pickle functions are **not** BC because the output type of `__getstate__` has changed.  We expect this to be acceptable as no user flows are actually broken (loading a v1 model with v2 code works), which is why we whitelist the failure.

Test plan (TODO finalize):

```
// adhoc testing of saving v1 and loading in v2: https://gist.github.com/vkuzo/f3616c5de1b3109cb2a1f504feed69be 

// test that loading models with v1 conv params format works and leads to the same numerics
python test/test_quantization.py TestSerialization.test_conv2d_graph
python test/test_quantization.py TestSerialization.test_conv2d_nobias_graph

// test that saving and loading models with v2 conv params format works and leads to same numerics
python test/test_quantization.py TestSerialization.test_conv2d_graph_v2
python test/test_quantization.py TestSerialization.test_conv2d_nobias_graph_v2

// TODO before land: 
// test numerics for a real model
// test legacy ONNX path
```

Note: this is a newer copy of https://github.com/pytorch/pytorch/pull/40003

Differential Revision: [D23347832](https://our.internmc.facebook.com/intern/diff/D23347832)

[ghstack-poisoned]

vkuzo committed:
Update on "Adding a version serialization type to ConvPackedParam"


This PR changes the format of `ConvPackedParam` in a nearly backwards-compatible way:
* a new format is introduced which has more flexibility and a lower on-disk size
* custom pickle functions are added to `ConvPackedParams` which know how to load the old format
* the custom pickle functions are **not** BC because the output type of `__getstate__` has changed.  We expect this to be acceptable as no user flows are actually broken (loading a v1 model with v2 code works), which is why we whitelist the failure.

Test plan (TODO finalize):

```
// adhoc testing of saving v1 and loading in v2: https://gist.github.com/vkuzo/f3616c5de1b3109cb2a1f504feed69be 

// test that loading models with v1 conv params format works and leads to the same numerics
python test/test_quantization.py TestSerialization.test_conv2d_graph
python test/test_quantization.py TestSerialization.test_conv2d_nobias_graph

// test that saving and loading models with v2 conv params format works and leads to same numerics
python test/test_quantization.py TestSerialization.test_conv2d_graph_v2
python test/test_quantization.py TestSerialization.test_conv2d_nobias_graph_v2

// test that numerics on quantized MobileNetV2 do not change:
// https://gist.github.com/vkuzo/d593385a88c30e22f46eba4248070c79

// TODO before land:
// figure out forward compatibility
```

Note: this is a newer copy of https://github.com/pytorch/pytorch/pull/40003

Differential Revision: [D23347832](https://our.internmc.facebook.com/intern/diff/D23347832)

[ghstack-poisoned]

vkuzo committed:
add stubs for new fields on "Adding a version serialization type to ConvPackedParam"


This PR changes the format of `ConvPackedParam` in a nearly backwards-compatible way:
* a new format is introduced which has more flexibility and a lower on-disk size
* custom pickle functions are added to `ConvPackedParams` which know how to load the old format
* the custom pickle functions are **not** BC because the output type of `__getstate__` has changed.  We expect this to be acceptable as no user flows are actually broken (loading a v1 model with v2 code works), which is why we whitelist the failure.

For a short forward compatibility period, we keep the `__getstate__` function with the old format,
while `__getstate__` knows about both old and new format.  The next PR in the stack will be landed
some period after this PR, and will switch `__getstate__` to the new format.

Note: the bc test cases were generated with the `v2` format, to allow manual testing of the `v2`
code in this PR.

Test plan:

```
// adhoc testing of saving v1 and loading in v2: https://gist.github.com/vkuzo/f3616c5de1b3109cb2a1f504feed69be 

// test that loading models with v1 conv params format works and leads to the same numerics
python test/test_quantization.py TestSerialization.test_conv2d_graph
python test/test_quantization.py TestSerialization.test_conv2d_nobias_graph

// test that saving and loading models with v2 conv params format works and leads to same numerics
python test/test_quantization.py TestSerialization.test_conv2d_graph_v2
python test/test_quantization.py TestSerialization.test_conv2d_nobias_graph_v2

// test that numerics on quantized MobileNetV2 do not change:
// https://gist.github.com/vkuzo/d593385a88c30e22f46eba4248070c79
```

Note: this is a newer copy of https://github.com/pytorch/pytorch/pull/40003

Differential Revision: [D23347832](https://our.internmc.facebook.com/intern/diff/D23347832)

[ghstack-poisoned]

vkuzo committed:
fix tabs on "Adding a version serialization type to ConvPackedParam"


This PR changes the format of `ConvPackedParam` in a nearly backwards-compatible way:
* a new format is introduced which has more flexibility and a lower on-disk size
* custom pickle functions are added to `ConvPackedParams` which know how to load the old format
* the custom pickle functions are **not** BC because the output type of `__getstate__` has changed.  We expect this to be acceptable as no user flows are actually broken (loading a v1 model with v2 code works), which is why we whitelist the failure.

For a short forward compatibility period, we keep the `__getstate__` function with the old format,
while `__getstate__` knows about both old and new format.  The next PR in the stack will be landed
some period after this PR, and will switch `__getstate__` to the new format.

Note: the bc test cases were generated with the `v2` format, to allow manual testing of the `v2`
code in this PR.

Test plan:

```
// adhoc testing of saving v1 and loading in v2: https://gist.github.com/vkuzo/f3616c5de1b3109cb2a1f504feed69be 

// test that loading models with v1 conv params format works and leads to the same numerics
python test/test_quantization.py TestSerialization.test_conv2d_graph
python test/test_quantization.py TestSerialization.test_conv2d_nobias_graph

// test that saving and loading models with v2 conv params format works and leads to same numerics
python test/test_quantization.py TestSerialization.test_conv2d_graph_v2
python test/test_quantization.py TestSerialization.test_conv2d_nobias_graph_v2

// test that numerics on quantized MobileNetV2 do not change:
// https://gist.github.com/vkuzo/d593385a88c30e22f46eba4248070c79
```

Note: this is a newer copy of https://github.com/pytorch/pytorch/pull/40003

Differential Revision: [D23347832](https://our.internmc.facebook.com/intern/diff/D23347832)

[ghstack-poisoned]

dreiss commented on aten/src/ATen/native/quantized/cpu/conv_serialization.h:
```diff
@@ -0,0 +1,292 @@
+ *      - dilation x kSpatialDim
+ *      - output_padding x kSpatialDim (unused)
+ *      - groups
+ *      - traspose (0 or 1, unused)
```
"transpose"

dreiss commented on aten/src/ATen/native/quantized/cpu/conv_serialization.h:
```diff
@@ -0,0 +1,292 @@
+    non_optional.emplace_back(std::move(weight));
+    optional.emplace_back(std::move(bias));
+
+    state = std::tie(version, non_optional, optional);
+
+  } else {
+    // version 2
+    state = v.to<ConvParamsSerializationType>();
+  }
+
+  return state;
```
I would suggest writing this as follows to reduce the risk of accidentally returning an empty result if this code changes in the future:

```

    return std::tie(version, non_optional, optional);

  } else if (version == 2) {
    // version 2
    return v.to<ConvParamsSerializationType>();

  } else {
    TORCH_INTERNAL_ASSERT(false, "Unexpected serialized qconv version: ", version);
  }
}
```

dreiss commented on aten/src/ATen/native/quantized/cpu/conv_serialization.h:
```diff
@@ -0,0 +1,292 @@
+
+  std::tie(version, non_optional, optional) = state;
+
+  // c10::optional<at::Tensor> bias;
```
Drop this comment?

dreiss commented on aten/src/ATen/native/quantized/cpu/conv_serialization.h:
```diff
@@ -0,0 +1,292 @@
+  std::tie(version, non_optional, optional) = state;
+
+  // c10::optional<at::Tensor> bias;
+  torch::List<int64_t> stride, padding, dilation;
+  int64_t groups;
```
Declare these closer to where they are used?

dreiss commented on aten/src/ATen/native/quantized/cpu/conv_serialization.h:
```diff
@@ -0,0 +1,292 @@
+  torch::List<int64_t> stride, padding, dilation;
+  int64_t groups;
+
+  at::Tensor conv_params_packed = non_optional[0];
+  at::Tensor weight = non_optional[1];
+  c10::optional<at::Tensor> bias = optional[0];
```
You're assuming version 2 here.  I would add an assert for that.

vkuzo replied:
hmm, this is enforced by the input type to the function, `ConvParamsSerializationType` is for v2+, and `ConvParamsSerializationTypeLegacy` is for v1.  Since looks like it's not 100% intuitive, sg on adding the assert.

dreiss commented on aten/src/ATen/native/quantized/cpu/conv_serialization.h:
```diff
@@ -0,0 +1,292 @@
+
+  // skip kSpatialDim
+  int idx = 1;
+  for (; idx < kSpatialDim + 1; ++idx) {
+    stride.emplace_back(conv_params_packed[idx].item<int64_t>());
+  }
```
Could write this (and the following loops as)

```
for (int i= 0; i < kSpatialDim; ++i) {
  stride.emplace_back(conv_params_packed[idx++].item<int64_t>());
}
```

(or `++idx` as a separate statement, or `i++, idx++` in the loop header).

That way, idx keeps counting up through all of the params you load, but you don't need to explicitly keep track of the exact value that idx is supposed to count up to.  You just make sure that each loop runs kSpatialDim times.

vkuzo replied:
sg.  tbh i'd prefer to increment idx by itself as well, for more clarity.

dreiss commented on aten/src/ATen/native/quantized/cpu/conv_serialization.h:
```diff
@@ -0,0 +1,292 @@
+  for (; idx < 4 * kSpatialDim + 1; ++idx) {
+    // do nothing
+  }
+  groups = conv_params_packed[idx].item<int64_t>();
```
I would change `idx` to `idx++` here so it's correct when you later want to add code to read `transpose` later.

Maybe consider asserting `idx == conv_params_packed.numel()` here.

dreiss commented on test/backward_compatibility/check_backward_compatibility.py:
```diff
@@ -69,6 +77,10 @@ def allow_listed(schema, allow_list):
+            if len(item) > 2:
+                # if arguments regex is present, use it
+                regexp_args = re.compile(item[2])
+                return True if regexp_args.search(str(schema)) else False
```
`return bool(regexp_args.search(str(schema)))`

dreiss commented on test/backward_compatibility/check_backward_compatibility.py:
```diff
@@ -18,6 +18,13 @@
 #
 # Allowlist entries can be removed after the date listed on them passes.
+#
+# Allowlist item format:
```
Maybe check with Lu about this change.

dreiss commented on aten/src/ATen/native/quantized/cpu/conv_serialization.h:
```diff
@@ -0,0 +1,292 @@
+  // skip kSpatialDim
+  int idx = 1;
+  for (; idx < kSpatialDim + 1; ++idx) {
+    stride.emplace_back(conv_params_packed[idx].item<int64_t>());
```
I'm mildly concerned about the cost of constructing a TensorImpl here, but that can be addressed later with no change to the on-disk format, so let's ignore it for now.  I'm planning to profile model load time soon anyway.

dreiss reviewed (approved):
Several style suggestions, but functionality looks solid.  Let's ship it!

vkuzo committed:
Update on "Adding a version serialization type to ConvPackedParam"


This PR changes the format of `ConvPackedParam` in a nearly backwards-compatible way:
* a new format is introduced which has more flexibility and a lower on-disk size
* custom pickle functions are added to `ConvPackedParams` which know how to load the old format
* the custom pickle functions are **not** BC because the output type of `__getstate__` has changed.  We expect this to be acceptable as no user flows are actually broken (loading a v1 model with v2 code works), which is why we whitelist the failure.

For a short forward compatibility period, we keep the `__getstate__` function with the old format,
while `__getstate__` knows about both old and new format.  The next PR in the stack will be landed
some period after this PR, and will switch `__getstate__` to the new format.

Note: the bc test cases were generated with the `v2` format, to allow manual testing of the `v2`
code in this PR.

Test plan:

```
// adhoc testing of saving v1 and loading in v2: https://gist.github.com/vkuzo/f3616c5de1b3109cb2a1f504feed69be 

// test that loading models with v1 conv params format works and leads to the same numerics
python test/test_quantization.py TestSerialization.test_conv2d_graph
python test/test_quantization.py TestSerialization.test_conv2d_nobias_graph

// test that saving and loading models with v2 conv params format works and leads to same numerics
python test/test_quantization.py TestSerialization.test_conv2d_graph_v2
python test/test_quantization.py TestSerialization.test_conv2d_nobias_graph_v2

// test that numerics on quantized MobileNetV2 do not change:
// https://gist.github.com/vkuzo/d593385a88c30e22f46eba4248070c79
```

Note: this is a newer copy of https://github.com/pytorch/pytorch/pull/40003

Differential Revision: [D23347832](https://our.internmc.facebook.com/intern/diff/D23347832)

[ghstack-poisoned]

vkuzo committed:
Update on "Adding a version serialization type to ConvPackedParam"


This PR changes the format of `ConvPackedParam` in a nearly backwards-compatible way:
* a new format is introduced which has more flexibility and a lower on-disk size
* custom pickle functions are added to `ConvPackedParams` which know how to load the old format
* the custom pickle functions are **not** BC because the output type of `__getstate__` has changed.  We expect this to be acceptable as no user flows are actually broken (loading a v1 model with v2 code works), which is why we whitelist the failure.

For a short forward compatibility period, we keep the `__getstate__` function with the old format,
while `__getstate__` knows about both old and new format.  The next PR in the stack will be landed
some period after this PR, and will switch `__getstate__` to the new format.

Note: the bc test cases were generated with the `v2` format, to allow manual testing of the `v2`
code in this PR.

Test plan:

```
// adhoc testing of saving v1 and loading in v2: https://gist.github.com/vkuzo/f3616c5de1b3109cb2a1f504feed69be 

// test that loading models with v1 conv params format works and leads to the same numerics
python test/test_quantization.py TestSerialization.test_conv2d_graph
python test/test_quantization.py TestSerialization.test_conv2d_nobias_graph

// test that saving and loading models with v2 conv params format works and leads to same numerics
python test/test_quantization.py TestSerialization.test_conv2d_graph_v2
python test/test_quantization.py TestSerialization.test_conv2d_nobias_graph_v2

// test that numerics on quantized MobileNetV2 do not change:
// https://gist.github.com/vkuzo/d593385a88c30e22f46eba4248070c79
```

Note: this is a newer copy of https://github.com/pytorch/pytorch/pull/40003

Differential Revision: [D23347832](https://our.internmc.facebook.com/intern/diff/D23347832)

[ghstack-poisoned]

dreiss commented on aten/src/ATen/native/quantized/cpu/conv_serialization.h:
```diff
@@ -229,31 +227,42 @@ c10::intrusive_ptr<ConvPackedParamsBase<kSpatialDim>> deserialize_conv(
+  // transpose is not implemented yet, so we skip the entry
+  idx++;
+  TORCH_INTERNAL_ASSERT(idx == conv_params_packed.numel(),
+      "Unexpected length of conv_params_packed, expected ",
+      conv_params_packed.numel(),
+      " got ",
+      idx);
```
I think the message here is reversed.  idx is what we expected.

vkuzo replied:
you're right, thanks!

vkuzo committed:
Update on "Adding a version serialization type to ConvPackedParam"


This PR changes the format of `ConvPackedParam` in a nearly backwards-compatible way:
* a new format is introduced which has more flexibility and a lower on-disk size
* custom pickle functions are added to `ConvPackedParams` which know how to load the old format
* the custom pickle functions are **not** BC because the output type of `__getstate__` has changed.  We expect this to be acceptable as no user flows are actually broken (loading a v1 model with v2 code works), which is why we whitelist the failure.

For a short forward compatibility period, we keep the `__getstate__` function with the old format,
while `__getstate__` knows about both old and new format.  The next PR in the stack will be landed
some period after this PR, and will switch `__getstate__` to the new format.

Note: the bc test cases were generated with the `v2` format, to allow manual testing of the `v2`
code in this PR.

Test plan:

```
// adhoc testing of saving v1 and loading in v2: https://gist.github.com/vkuzo/f3616c5de1b3109cb2a1f504feed69be 

// test that loading models with v1 conv params format works and leads to the same numerics
python test/test_quantization.py TestSerialization.test_conv2d_graph
python test/test_quantization.py TestSerialization.test_conv2d_nobias_graph

// test that saving and loading models with v2 conv params format works and leads to same numerics
python test/test_quantization.py TestSerialization.test_conv2d_graph_v2
python test/test_quantization.py TestSerialization.test_conv2d_nobias_graph_v2

// test that numerics on quantized MobileNetV2 do not change:
// https://gist.github.com/vkuzo/d593385a88c30e22f46eba4248070c79
```

Note: this is a newer copy of https://github.com/pytorch/pytorch/pull/40003

Differential Revision: [D23347832](https://our.internmc.facebook.com/intern/diff/D23347832)

[ghstack-poisoned]

vkuzo committed:
Update on "Adding a version serialization type to ConvPackedParam"


This PR changes the format of `ConvPackedParam` in a nearly backwards-compatible way:
* a new format is introduced which has more flexibility and a lower on-disk size
* custom pickle functions are added to `ConvPackedParams` which know how to load the old format
* the custom pickle functions are **not** BC because the output type of `__getstate__` has changed.  We expect this to be acceptable as no user flows are actually broken (loading a v1 model with v2 code works), which is why we whitelist the failure.

For a short forward compatibility period, we keep the `__getstate__` function with the old format,
while `__getstate__` knows about both old and new format.  The next PR in the stack will be landed
some period after this PR, and will switch `__getstate__` to the new format.

Note: the bc test cases were generated with the `v2` format, to allow manual testing of the `v2`
code in this PR.

Test plan:

```
// adhoc testing of saving v1 and loading in v2: https://gist.github.com/vkuzo/f3616c5de1b3109cb2a1f504feed69be 

// test that loading models with v1 conv params format works and leads to the same numerics
python test/test_quantization.py TestSerialization.test_conv2d_graph
python test/test_quantization.py TestSerialization.test_conv2d_nobias_graph

// test that saving and loading models with v2 conv params format works and leads to same numerics
python test/test_quantization.py TestSerialization.test_conv2d_graph_v2
python test/test_quantization.py TestSerialization.test_conv2d_nobias_graph_v2

// test that numerics on quantized MobileNetV2 do not change:
// https://gist.github.com/vkuzo/d593385a88c30e22f46eba4248070c79
```

Note: this is a newer copy of https://github.com/pytorch/pytorch/pull/40003

Differential Revision: [D23347832](https://our.internmc.facebook.com/intern/diff/D23347832)

[ghstack-poisoned]

vkuzo committed:
Update on "Adding a version serialization type to ConvPackedParam"


This PR changes the format of `ConvPackedParam` in a nearly backwards-compatible way:
* a new format is introduced which has more flexibility and a lower on-disk size
* custom pickle functions are added to `ConvPackedParams` which know how to load the old format
* the custom pickle functions are **not** BC because the output type of `__getstate__` has changed.  We expect this to be acceptable as no user flows are actually broken (loading a v1 model with v2 code works), which is why we whitelist the failure.

For a short forward compatibility period, we keep the `__getstate__` function with the old format,
while `__getstate__` knows about both old and new format.  The next PR in the stack will be landed
some period after this PR, and will switch `__getstate__` to the new format.

Note: the bc test cases were generated with the `v2` format, to allow manual testing of the `v2`
code in this PR.

Test plan:

```
// adhoc testing of saving v1 and loading in v2: https://gist.github.com/vkuzo/f3616c5de1b3109cb2a1f504feed69be 

// test that loading models with v1 conv params format works and leads to the same numerics
python test/test_quantization.py TestSerialization.test_conv2d_graph
python test/test_quantization.py TestSerialization.test_conv2d_nobias_graph

// test that saving and loading models with v2 conv params format works and leads to same numerics
python test/test_quantization.py TestSerialization.test_conv2d_graph_v2
python test/test_quantization.py TestSerialization.test_conv2d_nobias_graph_v2

// test that numerics on quantized MobileNetV2 do not change:
// https://gist.github.com/vkuzo/d593385a88c30e22f46eba4248070c79
```

Note: this is a newer copy of https://github.com/pytorch/pytorch/pull/40003

Differential Revision: [D23347832](https://our.internmc.facebook.com/intern/diff/D23347832)

[ghstack-poisoned]
